apps/wappush/js/wappush.js

```diff
@@ -282,49 +266,127 @@
+    if (displayedSiMessage &&
+      displayedSiMessage.getUniqueId() === message.getUniqueId()) {
+      wpm_close();
+    }
+
+    // Remove the notification associated to the deleted message
+    wpm_clearNotifications(+message.getUniqueId());
+  }
+
+  /**
+   * Prints a debug statement when a message has been discarded, no further
+   * action required.
+   */
+  function wpm_onDiscard() {
+    DUMP('The message was discarded');
+    wpm_finish();
   }
 
   /**
    * Retrieves a WAP Push message from the database and displays it
    *
-   * @param {String} timestamp The message timestamp as a string.
+   * @param {String} id The message id as a string.
    * @return {Promise} A promise that resolves once the message has been
    *                   displayed and the associated notification cleared.
    */
-  function wpm_displayWapPushMessage(timestamp) {
-    DUMP('Displaying message ' + timestamp);
+  function wpm_displayWapPushMessage(id) {
+    DUMP('Displaying message ' + id);
 
     // Clear the close timer as the application will soon become visible
     app.launch();
     window.clearTimeout(closeTimeout);
     closeTimeout = null;
 
-    return ParsedMessage.load(timestamp).then(
+    return ParsedMessage.load(id).then(
       function wpm_loadResolved(message) {
-        if (message) {
-          switch (message.type) {
-            case 'text/vnd.wap.si':
-            case 'text/vnd.wap.sl':
+        switch (message.type) {
+          case 'text/vnd.wap.si':
+            displayedSiMessage = message;
+            CpScreenHelper.hide();
+
+            if (message.isExpired()) {
+              // Notify the user that the message has expired
+              SiSlScreenHelper.populateScreen();
+            } else {
               SiSlScreenHelper.populateScreen(message);
-              return wpm_clearNotifications(timestamp);
-            case 'text/vnd.wap.connectivity-xml':
-              CpScreenHelper.showConfirmInstallationDialog(message);
-              break;
-          }
-        } else {
-          // Notify the user that the message has expired
-          SiSlScreenHelper.populateScreen();
-          return wpm_clearNotifications(timestamp);
+            }
+
+            return wpm_clearNotifications(id);
+
+          case 'text/vnd.wap.sl':
+            displayedSiMessage = null;
+            CpScreenHelper.hide();
+            SiSlScreenHelper.populateScreen(message);
+            return wpm_clearNotifications(id);
+
+          case 'text/vnd.wap.connectivity-xml':
+            displayedSiMessage = null;
+            SiSlScreenHelper.hide();
+            CpScreenHelper.showConfirmInstallationDialog(message);
+            return Promise.resolve();
         }
       }
     ).catch(function(error) {
-      console.error('Could not retrieve the message: ' + error.name + '\n');
+      var message = error.message || 'Unknown error';
+      console.error('Could not retrieve the message: ' + message + '\n');
```

Comment:
nit: you don't need '\n', and thus you can just separate the prefix from the message with a comma